File: npm-package/android/src/main/java/expo/modules/walletcore/chains/SolanaSigning.kt
package expo.modules.walletcore.chains

import com.google.protobuf.ByteString
import expo.modules.walletcore.WalletCoreError
import java.util.Arrays
import wallet.core.java.AnySigner
import wallet.core.jni.CoinType
import wallet.core.jni.HDWallet
import wallet.core.jni.proto.Solana

/**
 * Solana signing helpers. Called from WalletCoreModule.definition()'s
 * solanaSignTransfer AsyncFunction. v0.2.0 binds the basic SOL transfer
 * to validate the multi-chain package layout; additional instructions
 * follow the same pattern.
 */
internal object SolanaSigning {

    fun signTransfer(
        tx: Map<String, Any?>,
        mnemonic: String,
        derivationPath: String,
    ): String {
        val wallet = try {
            HDWallet(mnemonic, "")
        } catch (e: Exception) {
            throw WalletCoreError.InvalidMnemonic()
        }
        val privateKey = wallet.getKey(CoinType.SOLANA, derivationPath)

        // Memory hygiene — same W-06 pattern as EvmSigning.kt: zero-fill the
        // extracted key bytes on every exit path after AnySigner.sign returns.
        val pkBytes = privateKey.data()
        try {
            val recipient = (tx["recipient"] as? String)
                ?: throw WalletCoreError.MissingField("recipient")
            val recentBlockhash = (tx["recentBlockhash"] as? String)
                ?: throw WalletCoreError.MissingField("recentBlockhash")

            // Java's signed long can hold the full uint64 bit pattern when parsed
            // via parseUnsignedLong; the protobuf wire format treats it as uint64.
            val lamports: Long = when (val raw = tx["lamports"]) {
                null -> throw WalletCoreError.MissingField("lamports")
                is String -> try {
                    java.lang.Long.parseUnsignedLong(raw)
                } catch (e: NumberFormatException) {
                    throw WalletCoreError.InvalidField("lamports", "not a valid uint64 ($raw)")
                }
                is Number -> raw.toLong()
                else -> throw WalletCoreError.InvalidField("lamports", "expected uint64 string or number")
            }

            val transferBuilder = Solana.Transfer.newBuilder()
                .setRecipient(recipient)
                .setValue(lamports)
            (tx["memo"] as? String)?.let { transferBuilder.memo = it }

            val input = Solana.SigningInput.newBuilder()
                .setPrivateKey(ByteString.copyFrom(pkBytes))
                .setRecentBlockhash(recentBlockhash)
                .setTransferTransaction(transferBuilder.build())
                .build()

            // See EvmSigning.kt for the AnySigner.sign<T>(MessageLite, CoinType, Parser<T>) generic.
            val output: Solana.SigningOutput = AnySigner.sign(
                input,
                CoinType.SOLANA,
                Solana.SigningOutput.parser(),
            )
            return output.encoded
        } finally {
            Arrays.fill(pkBytes, 0.toByte())
        }
    }
}

File: npm-package/ios/chains/SolanaSigning.swift
import Foundation
import SwiftProtobuf

// Solana signing helpers. Called from WalletCoreModule.definition()'s
// `solanaSignTransfer` AsyncFunction. wallet-core's Solana support covers
// many instruction types; v0.2.0 binds only the basic transfer to validate
// the multi-chain package layout. Additional instructions (token transfer,
// stake, vote, etc.) follow the same pattern.

enum SolanaSigning {
  /// Sign a SOL transfer. Returns the wallet-core SigningOutput.encoded string
  /// (an encoded Solana transaction, ready for `sendTransaction` RPC).
  static func signTransfer(
    tx: [String: Any],
    mnemonic: String,
    derivationPath: String
  ) throws -> String {
    guard let wallet = HDWallet(mnemonic: mnemonic, passphrase: "") else {
      throw WalletCoreError.invalidMnemonic
    }
    let privateKey = wallet.getKey(coin: .solana, derivationPath: derivationPath)

    // Memory hygiene — same W-06 pattern as EvmSigning.swift: zero-fill the
    // extracted key bytes on every exit path after AnySigner.sign returns.
    var pkData = privateKey.data
    var input = TW_Solana_Proto_SigningInput()
    defer {
      pkData.resetBytes(in: 0..<pkData.count)
      input.privateKey = Data()
    }

    guard let recipient = tx["recipient"] as? String else {
      throw WalletCoreError.missingField("recipient")
    }
    guard let lamportsValue = tx["lamports"] else {
      throw WalletCoreError.missingField("lamports")
    }
    let lamports: UInt64
    switch lamportsValue {
    case let s as String:
      guard let parsed = UInt64(s) else {
        throw WalletCoreError.invalidField("lamports", "not a valid uint64 (\(s))")
      }
      lamports = parsed
    case let n as NSNumber:
      lamports = n.uint64Value
    default:
      throw WalletCoreError.invalidField("lamports", "expected uint64 string or number")
    }
    guard let recentBlockhash = tx["recentBlockhash"] as? String else {
      throw WalletCoreError.missingField("recentBlockhash")
    }

    input.privateKey = pkData
    input.recentBlockhash = recentBlockhash
    input.transferTransaction = TW_Solana_Proto_Transfer.with {
      $0.recipient = recipient
      $0.value = lamports
      if let memo = tx["memo"] as? String {
        $0.memo = memo
      }
    }

    // See EvmSigning.swift for the AnySigner.sign<Output>(input:coin:) generic.
    let output: TW_Solana_Proto_SigningOutput =
        AnySigner.sign(input: input, coin: .solana)
    return output.encoded
  }
}
